onnxruntime/core/optimizer/selectors_actions/selector_action_transformer.cc

```diff
@@ -120,86 +151,126 @@ Status SelectorActionTransformer::MatchAndProcess(Graph& graph, const GraphViewe
-  GraphViewer graph_viewer(graph);
+#if defined(ORT_ENABLE_ORT_FORMAT_RUNTIME_GRAPH_OPTIMIZATION)
 
-  for (auto index : graph_viewer.GetNodesInTopologicalOrder()) {
-    auto* node = graph.GetNode(index);
-    if (node == nullptr) {
-      continue;  // was removed by this transformer
+static Status RegisterProducedNodesWithGraph(NodeIndex pre_action_max_num_nodes, NodeIndex post_action_max_num_nodes,
+                                             const RuntimeOptimizationRecord& record,
+                                             Graph& graph) {
+  assert(post_action_max_num_nodes >= pre_action_max_num_nodes);
+
+  const auto num_new_node_indices = post_action_max_num_nodes - pre_action_max_num_nodes;
+
+  auto produced_node_it = record.produced_nodes.begin();
+  const auto produced_nodes_end = record.produced_nodes.end();
+
+  std::unordered_map<NodeIndex, HashValue> node_index_to_kernel_def_hash{};
+
+  for (NodeIndex i = 0; i < num_new_node_indices; ++i) {
+    const NodeIndex new_node_idx = pre_action_max_num_nodes + i;
+    const auto* new_node = graph.GetNode(new_node_idx);
+
+    if (!new_node) continue;
+
+    ORT_RETURN_IF(produced_node_it == produced_nodes_end,
+                  "Not enough produced nodes in the runtime optimization record.");
+
+    node_index_to_kernel_def_hash.emplace(new_node_idx, produced_node_it->kernel_def_hash);
+
+    ++produced_node_it;
+  }
+
+  ORT_RETURN_IF(produced_node_it != produced_nodes_end, "Too many produced nodes in the runtime optimization record.");
+
+  graph.MutableRuntimeOptimizationReplayCtx().produced_node_index_to_kernel_def_hash.merge(
+      node_index_to_kernel_def_hash);
+
+  return Status::OK();
+}
+
+Status SelectorActionTransformer::ApplyFromRuntimeOptimizations(
+    Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  for (auto& node : graph.Nodes()) {
+    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+  }
+
+  const auto records = graph.MutableRuntimeOptimizations().RemoveRecordsForOptimizer(Name());
+  for (const auto& record : records) {
+    const auto* action = selectors_and_actions_.LookUpAction(record.action_id);
+    if (!action) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Missing action ", record.action_id, " for transformer ", Name());
     }
 
-    ORT_RETURN_IF_ERROR(Recurse(*node, modified, graph_level, logger));
+    NodesToOptimize nodes_to_optimize{graph, record.nodes_to_optimize_indices};
 
-#if !defined(ORT_MINIMAL_BUILD)
-    // TODO: use GraphTransformer::GetCompatibleExecutionProviders if we need something more flexible
-    if (node->GetExecutionProviderType() == kCpuExecutionProvider) {
-      ORT_RETURN_IF_ERROR(MatchAndProcess(graph, graph_viewer, *node, modified, logger));
+    if (!nodes_to_optimize.IsValid()) {
+      continue;
     }
-#else
-    ORT_RETURN_IF_ERROR(ApplySaved(graph, modified, logger));
-#endif
+
```

Comment:
Would `AllExist` be more meaningful than `IsValid`?

Reply:
I think `IsValid` still makes sense - if valid, then we can proceed